### ISSUE TITLE
[feature request] [PyTorch] Dynamic Samplers.

### ISSUE THREAD

**Issue by kris-singh** [pytorch/pytorch#7359]
I am trying to implement Dynamic Samplers in Pytorch. This seems impossible with the existing train_loader api's. As the weighted the samplers only allow weights to be set once. I was thinking that weights could change based on the loss value for each and individual data point. This can be done without train_loader API but I think this a feature that could be included in a future release. 

Something as simple as Loss Based sampling can't be done right now.

**Comment by SsnL** [comment]
I think for something like this you should write your own `torch.data.utils.sampler.Sampler` subclass.

**Comment by kris-singh** [comment]
@SsnL Actually you can't since the sampler functions accept no arguments in the __iter__ function. You can't pass the update loss values after training for one epoch has been done.

**Comment by SsnL** [comment]
Oh so you want it to change during training, right? Well you can still do it via a subclass by just adding an extra method on it to update. However, you should pay attention to only use `num_workers=0`.

Considering that samplers are designed to work with multiprocessing data loading, I don't think this is very doable generally.

**Comment by apaszke** [comment]
It will work even with multiple workers. The sampler never leaves the main process

**Comment by SsnL** [comment]
@apaszke I know that. But the batch indices are generated at the previous `__next__` (not current), which is before the users can update the weight.

**Comment by apaszke** [comment]
Right. It all depends on what kind of consistency you're expecting. If it's some kind of slowly moving guidance based on successive losses, it might be ok to have a delay of a few batches.

**Comment by kris-singh** [comment]
I decided to implement a random sampler by myself. As I am testing my own sampling strategy against random sampling and other sampling mechanisms. I tried to implement by random sampling strategy based on RandomSampler method. But the results are vastly different I am getting a test accuracy of 15%. But the expected test accuracy is 95%. I can’t figure out what I am doing wrong.
```
# Read the data
dataset = datasets.MNIST('./mnist', train=True, download=False,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ]))
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=1000, shuffle=False)
data, target = dataset.train_data, dataset.train_labels
data = data.unsqueeze(1).type(torch.FloatTensor)

# Intialize a model
model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

# Random Sampler
def temp(data, batch_size):
    for i in range(0 , data.shape[0] // batch_size):
        sampled = torch.randperm(len(data))[0:batch_size]
        yield i, sampled

# Training Network
for epoch in range(50):
    for batch_idx, i in temp(50):
        model.train(data)
        optimizer.zero_grad()
        d=data[i].to(device)
        t=target[i].to(device)
        output,_ = model(d)
        loss = F.nll_loss(output, t)
        loss.backward()
        optimizer.step()
        if  batch_idx % 25 == 0:
            print('Train Epoch:  [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                batch_idx * len(d), len(data),
                100. * batch_idx / len(data), loss.item()))
    test(model, device, test_loader)
    ```

**Comment by kris-singh** [comment]
@apaszke @SsnL  I think I figured out a way to loss based sampling. Here is the code.
```python
from __future__ import division
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data.sampler import Sampler
# np.random.seed(1337)
# torch.manual_seed(1337)
# torch.cuda.manual_seed_all(1337)

device = torch.device ("cuda:2" if torch.cuda.is_available () else "cpu")


class RandomSampler(Sampler):
    r"""Samples elements randomly, without replacement.
    Arguments:
        data_source (Dataset): dataset to sample from
    """

    def __init__(self, model, data_source, train_data, train_target, batch_size):
        self.model = model
        self.data_source = data_source
        self.batch_size = batch_size
        self.data = train_data
        self.data = torch.unsqueeze(self.data, 1)
        self.data = self.data.type(torch.cuda.FloatTensor)
        self.target = train_target

    def get_scores(self):
        output, feat =self.model.forward(self.data)
        criterion = nn.CrossEntropyLoss (reduce=False)
        loss = criterion(output, self.target)
        return loss

    def __iter__(self):
        num_batches = len (self.data_source) // self.batch_size
        while num_batches > 0:
            scores = self.get_scores ()
            sampled = []
            while len(sampled) < self.batch_size:
                if len(sampled) == 0:
                    sampled.append(torch.argmax(scores))
                else:
                    sampled.append(torch.argmax(scores))
            yield sampled
            num_batches -=1

    def __len__(self):
        return len(self.data_source)


class BatchSampler(Sampler):

    def __init__(self, sampler, batch_size, drop_last):
        self.sampler = sampler
        self.batch_size = batch_size
        self.drop_last = drop_last

    def __iter__(self):
        batch = []
        for _, idx in enumerate(iter(self.sampler)):
            batch = idx
            yield batch

        if len(batch) > 0 and not self.drop_last:
            yield batch

    def __len__(self):
        return len(self.sampler) // self.batch_size


# Create the Network
class Net (nn.Module):
    def __init__(self):
        super (Net, self).__init__ ()
        self.conv1 = nn.Conv2d (1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d (10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d ()
        self.fc1 = nn.Linear (320, 50)
        self.fc2 = nn.Linear (50, 10)

    def forward(self, x):
        x = F.relu (F.max_pool2d (self.conv1 (x), 2))
        x = F.relu (F.max_pool2d (self.conv2_drop (self.conv2 (x)), 2))
        x = x.view (-1, 320)
        x = F.relu (self.fc1 (x))
        feat_x = x
        x = F.dropout (x, training=self.training)
        x = self.fc2 (x)
        return F.log_softmax (x, dim=1), feat_x



def train_normal(model, device, train_loader, optimizer, epoch):
    model.train ()
    for batch_idx, (data, target) in enumerate (train_loader):
        data = data.to (device)
        target = target.to (device)
        optimizer.zero_grad ()
        output, feat = model (data)
        loss = F.nll_loss (output, target)
        loss.backward ()
        optimizer.step ()
        if batch_idx % 100 == 0:
            print ('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format (
                epoch, batch_idx * len (data), len (train_loader.dataset),
                       100. * batch_idx / len (train_loader), loss.item ()))


def test(model, device, test_loader, epoch):
    model.eval ()
    test_loss = 0
    correct = 0
    with torch.no_grad ():
        for data, target in test_loader:
            data = data.to (device)
            target = target.to (device)
            output, _ = model (data)
            test_loss += F.nll_loss (output, target, size_average=False).item ()  # sum up batch loss
            pred = output.max (1, keepdim=True)[1]  # get the index of the max log-probability
            correct += pred.eq (target.view_as (pred)).sum ().item ()
    test_loss /= len (test_loader.dataset)
    print ('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format (
        test_loss, correct, len (test_loader.dataset),
        100. * correct / len (test_loader.dataset)))


def main():
    device = torch.device ("cuda:0" if torch.cuda.is_available () else "cpu")
    train_batch_size = 64
    test_batch_size = 1000
    learning_rate = 0.01
    momentum = 0.5
    epochs = 10
    # Read Dataset
    dataset = datasets.MNIST ("~/MNIST_data/", train=True, download=False,
                              transform=transforms.Compose ([
                                  transforms.ToTensor (),
                                  transforms.Normalize ((0.1307,), (0.3081,))
                              ]))

    test_dataset = datasets.MNIST ("~/MNIST_data/", train=False, download=False,
                                   transform=transforms.Compose ([
                                       transforms.ToTensor (),
                                       transforms.Normalize ((0.1307,), (0.3081,))
                                   ]))

    # Training loop
    model = Net ().to (device)
    train_data = dataset.train_data
    train_target = dataset.train_labels
    train_target = train_target.to(device)
    train_data = train_data.to(device)
    sampler = RandomSampler (model, dataset, train_data, train_target, train_batch_size)
    batch_sampler = BatchSampler (sampler, train_batch_size, True)
    train_loader = torch.utils.data.DataLoader (dataset, batch_sampler=batch_sampler, num_workers=0)
    test_loader = torch.utils.data.DataLoader (test_dataset, batch_size=test_batch_size, shuffle=False, num_workers=0)
    optimizer = optim.SGD (model.parameters (), lr=learning_rate, momentum=momentum)
    for i in range (epochs):
        train_normal (model, device, train_loader, optimizer, i)
        test (model, device, test_loader, i)

main ()
```
Please let me know if there are any comments.

**Comment by askerlee** [comment]
I wrote a WeightedSubsetRandomSampler which simply combines WeightedRandomSampler and SubsetRandomSampler. You can reassign the weights for each instances after one epoch is finished:

```
class WeightedSubsetRandomSampler(Sampler):
    r"""Samples elements from a given list of indices with given probabilities (weights), with replacement.

    Arguments:
        weights (sequence)   : a sequence of weights, not necessary summing up to one
        num_samples (int): number of samples to draw
    """

    def __init__(self, indices, weights, num_samples=0):
        if not isinstance(num_samples, _int_classes) or isinstance(num_samples, bool):
            raise ValueError("num_samples should be a non-negative integeral "
                             "value, but got num_samples={}".format(num_samples))
        self.indices = indices
        weights = [ weights[i] for i in self.indices ]
        self.weights = torch.tensor(weights, dtype=torch.double)
        if num_samples == 0:
            self.num_samples = len(self.weights)
        else:
            self.num_samples = num_samples
        self.replacement = True

    def __iter__(self):
        return (self.indices[i] for i in torch.multinomial(self.weights, self.num_samples, self.replacement))

    def __len__(self):
        return self.num_samples

class MyDataset(Dataset):
        ...
     # return data idx here, so that the corresponding weight can be updated based on the training loss.
    def __getitem__(self, idx):
        ...
        return image, label, idx

sample_weights = np.ones(len(train_labels))
for epoch in range(0, total_epochs):
        # tr_indices indexes a subset of train_labels
        train_sampler = WeightedSubsetRandomSampler(tr_indices, sample_weights)
        train_loader = torch.utils.data.DataLoader(train_set, batch_size=batch_size, sampler=train_sampler, num_workers=num_workers)

        for batch_i, (data, batch_truth, data_indices) in enumerate(train_loader):
              ...
              batch_loss = criterion_individual(scores, batch_truth)
              # update sample weights to be the loss, so that harder samples have larger chances to be drawn in the next epoch
              sample_weights[data_indices] = batch_loss.detach().cpu().numpy()
```

**Comment by stgzr** [comment]
By the way, I wonder if there is a way to change indices at each iteration? Because I want to change samples' weights during training, so each iteration the sample weights is different. Does `batch_sampler` help?
